### PR TITLE
With local openidc handler

### PR DESCRIPTION
**Pull request by cviecco**

### CODE REVIEW

**Comment by rgooch** [comment]
Can you add a high-level description of the changes? The commit logs don't reveal much...

File: broker/httpd/auth.go [hunk 1]
```diff
@@ -23,61 +55,234 @@ func randomStringGeneration() (string, error) {
...
+		username = userInfo.Email
+	}
+	return username
+}
 
-	userCookie := http.Cookie{Name: authCookieName, Value: randomString, Path: "/", Expires: expires, HttpOnly: true, Secure: true}
+func (s *Server) getBytesFromSuccessfullPost(url string, data url.Values) ([]byte, error) {
+	response, err := s.netClient.PostForm(url, data)
+	if err != nil {
+		s.logger.Printf("err=%s", err)
```
Comment (rgooch) [inline broker/httpd/auth.go]: Please replace with s.logger.Println(err)

File: broker/httpd/auth.go [hunk 2]
```diff
@@ -23,61 +55,234 @@ func randomStringGeneration() (string, error) {
...
+		return nil, err
+	}
+	defer response.Body.Close()
 
-	http.SetCookie(w, &userCookie)
+	responseBody, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		s.logger.Printf("err=%s", err)
```
Comment (rgooch) [inline broker/httpd/auth.go]: Please replace with s.logger.Println(err)

File: broker/httpd/auth.go [hunk 3]
```diff
@@ -23,61 +55,234 @@ func randomStringGeneration() (string, error) {
...
+	tok, err := jwt.ParseSigned(serializedState)
+	if err != nil {
+		return inboundJWT, err
+	}
+	if err := s.JWTClaims(tok, &inboundJWT); err != nil {
+		s.logger.Printf("error parsing claims err=%s", err)
+		return inboundJWT, err
+	}
+	// At this point we know the signature is valid, but now we must
+	//validate the contents of the jtw token
```
Comment (rgooch) [inline broker/httpd/auth.go]: Please add a space after the "//".
s/jwt/JWT/

File: broker/httpd/auth.go [hunk 4]
```diff
@@ -23,61 +55,234 @@ func randomStringGeneration() (string, error) {
...
+	}
+	authCode := r.URL.Query().Get("code")
+	if len(authCode) < 1 {
+		s.logger.Println("null code")
+		http.Error(w, "null code", http.StatusUnauthorized)
+		return
+	}
+	inboundJWT, err := s.getVerifyReturnStateJWT(r)
+	if err != nil {
+		s.logger.Printf("error processing state err=%s", err)
```
Comment (rgooch) [inline broker/httpd/auth.go]: Formally, there should be a trailing "\n". Same comment applies elsewhere.

File: broker/staticconfiguration/impl.go [hunk 5]
```diff
@@ -1,13 +1,39 @@
...
 import (
+	"bufio"
 	"errors"
 	"os"
 
 	"github.com/Symantec/cloud-gate/lib/constants"
 	"gopkg.in/yaml.v2"
 )
 
+func getClusterSecretsFile(culterSecretsFilename string) ([]string, error) {
```
Comment (rgooch) [inline broker/staticconfiguration/impl.go]: s/culter/cluster/
Same below.

File: broker/staticconfiguration/impl.go [hunk 6]
```diff
@@ -38,6 +64,23 @@ func LoadVerifyConfigFile(configFilename string) (*StaticConfiguration, error) {
 		config.Base.AccountConfigurationCheckInterval =
 			constants.DefaultAccountConfigurationCheckInterval
 	}
+	//verify oauth2 setup
```
Comment (rgooch) [inline broker/staticconfiguration/impl.go]: Please add space after "//".

File: broker/staticconfiguration/impl.go [hunk 7]
```diff
@@ -38,6 +64,23 @@ func LoadVerifyConfigFile(configFilename string) (*StaticConfiguration, error) {
...
+	if len(config.OpenID.AuthURL) < 1 ||
+		len(config.OpenID.TokenURL) < 1 ||
+		len(config.OpenID.UserinfoURL) < 1 ||
+		len(config.OpenID.Scopes) < 1 ||
+		len(config.OpenID.ClientID) < 1 {
+		return nil, errors.New("invalid openid config")
+	}
+
+	//verify shared secrets
```
Comment (rgooch) [inline broker/staticconfiguration/impl.go]: Please add space after "//".

File: broker/httpd/auth.go [hunk 8]
```diff
@@ -23,61 +55,234 @@ func randomStringGeneration() (string, error) {
...
-	return destinationPath
+func (s *Server) generateValidStateString(r *http.Request) (string, error) {
+	key := []byte(s.staticConfig.Base.SharedSecrets[0])
+	sig, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.HS256, Key: key}, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		s.logger.Printf("err=%s", err)
+		return "", err
+	}
+	issuer := "cloud-gate"
+	subject := "state:" + constants.RedirCookieName
```
[the remark this answers is not in the document]
Comment (cviecco) [reply broker/httpd/auth.go]: I want to ensure they are consistent, in the struct fields they appear in more than one place.

File: broker/httpd/auth.go [hunk 9]
```diff
@@ -23,61 +54,230 @@ func randomStringGeneration() (string, error) {
...
+	buf.WriteString(v.Encode())
+	return buf.String()
+}
 
-	return destinationPath
+func (s *Server) generateValidStateString(r *http.Request) (string, error) {
+	key := []byte(s.staticConfig.Base.SharedSecrets[0])
+	sig, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.HS256, Key: key}, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		s.logger.Debugf(1, "New jose signer error err=%s", err)
```
Comment (rgooch) [inline broker/httpd/auth.go]: I'm still seeing multiple instances of this "err=%s" format. Please use this form instead:
`s.logger.Debugf(1, "New jose signer error: %s", err)`

File: broker/httpd/auth.go [hunk 10]
```diff
@@ -23,61 +54,230 @@ func randomStringGeneration() (string, error) {
...
+		s.logger.Printf("token type invalid token=%s", string(tokenRespBody))
+		http.Error(w, "invalid accessToken ", http.StatusInternalServerError)
+		return
+	}
+
+	// Now we use the access_token (from token exchange) to get userinfo
+	userInfoRespBody, err := s.getBytesFromSuccessfullPost(s.staticConfig.OpenID.UserinfoURL,
+		url.Values{"access_token": {oauth2AccessToken.AccessToken}})
+	if err != nil {
+		s.logger.Printf("err=%s", err)
```
Comment (rgooch) [inline broker/httpd/auth.go]: Please replace with: `s.logger.Println(err)`

**Pull request merged**